core/processor/abci.go

```diff
@@ -641,98 +656,93 @@ func (app *App) Info(_ tmtypes.RequestInfo) tmtypes.ResponseInfo {
 	if !bytes.Equal(deserializedSnapshot.Hash, req.AppHash) {
 		app.log.Error("The hashes from the request and the deserialized snapshot mismatch",
 			logging.String("deserialized-hash", hex.EncodeToString(deserializedSnapshot.Hash)),
 			logging.String("request-hash", hex.EncodeToString(req.AppHash)))
-		return tmtypes.ResponseOfferSnapshot{
+		return &tmtypes.ResponseOfferSnapshot{
 			Result: tmtypes.ResponseOfferSnapshot_REJECT,
-		}
+		}, fmt.Errorf("hash mismatch")
 	}
 
-	return app.snapshotEngine.ReceiveSnapshot(deserializedSnapshot)
+	res := app.snapshotEngine.ReceiveSnapshot(deserializedSnapshot)
+	return &res, nil
 }
 
-func (app *App) ApplySnapshotChunk(ctx context.Context, req tmtypes.RequestApplySnapshotChunk) tmtypes.ResponseApplySnapshotChunk {
+func (app *App) ApplySnapshotChunk(ctx context.Context, req *tmtypes.RequestApplySnapshotChunk) (*tmtypes.ResponseApplySnapshotChunk, error) {
 	app.log.Debug("ABCI service ApplySnapshotChunk start")
 
 	if app.snapshotEngine.HasRestoredStateAlready() {
 		app.log.Warn("The snapshot engine aborted the snapshot chunk from state-sync since the state has already been restored")
-		return tmtypes.ResponseApplySnapshotChunk{
+		return &tmtypes.ResponseApplySnapshotChunk{
 			Result: tmtypes.ResponseApplySnapshotChunk_ABORT,
-		}
+		}, nil // ???
 	}
-
 	chunk := &types.RawChunk{
 		Nr:   req.Index,
 		Data: req.Chunk,
 	}
 
-	return app.snapshotEngine.ReceiveSnapshotChunk(ctx, chunk, req.Sender)
+	res := app.snapshotEngine.ReceiveSnapshotChunk(ctx, chunk, req.Sender)
+	return &res, nil
 }
 
-func (app *App) LoadSnapshotChunk(req tmtypes.RequestLoadSnapshotChunk) tmtypes.ResponseLoadSnapshotChunk {
+func (app *App) LoadSnapshotChunk(_ context.Context, req *tmtypes.RequestLoadSnapshotChunk) (*tmtypes.ResponseLoadSnapshotChunk, error) {
 	app.log.Debug("ABCI service LoadSnapshotChunk start")
-
-	rawChunk, err := app.snapshotEngine.RetrieveSnapshotChunk(req.Height, req.Format, req.Chunk)
+	raw, err := app.snapshotEngine.RetrieveSnapshotChunk(req.Height, req.Format, req.Chunk)
 	if err != nil {
-		app.log.Error("Could not load a snapshot chunk from snapshot engine",
-			logging.Uint64("height", req.Height),
-			logging.Error(err),
-		)
-		return tmtypes.ResponseLoadSnapshotChunk{}
-	}
-
-	return tmtypes.ResponseLoadSnapshotChunk{
-		Chunk: rawChunk.Data,
+		app.log.Error("failed to load snapshot chunk", logging.Error(err), logging.Uint64("height", req.Height))
+		return &tmtypes.ResponseLoadSnapshotChunk{}, err
 	}
+	return &tmtypes.ResponseLoadSnapshotChunk{
+		Chunk: raw.Data,
+	}, nil
 }
 
-func (app *App) OnInitChain(req tmtypes.RequestInitChain) tmtypes.ResponseInitChain {
+func (app *App) OnInitChain(req *tmtypes.RequestInitChain) (*tmtypes.ResponseInitChain, error) {
 	app.log.Debug("ABCI service InitChain start")
-	hash := hex.EncodeToString(vgcrypto.Hash([]byte(req.ChainId)))
+	hash := hex.EncodeToString(vgcrypto.Hash(req.AppStateBytes))
```

Comment:
The app-state bytes is not deterministic because the app-state contains maps, so this needs to stay as `vgcrypto.Hash([]byte(req.ChainId)` otherwise the event file diff will complain.